File: Samples/ApiGateway/Python/apigateway-lambda/lab_stack.py
from aws_cdk import (
    # Duration,
    Stack,
    # aws_sqs as sqs,
    aws_apigateway as apigateway,
    aws_lambda as lambda_, CfnOutput,
)
from aws_cdk.aws_apigateway import RestApi, Model, LambdaIntegration
from constructs import Construct


class LabStack(Stack):

    def __init__(self, scope: Construct, construct_id: str, rest_api_id: str, request_model: Model,
                 rest_api_root_resource_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        # https://docs.aws.amazon.com/cdk/api/v2/python/aws_cdk.aws_apigateway/RestApi.html#aws_cdk.aws_apigateway.RestApi.from_rest_api_id
        imported_api = RestApi.from_rest_api_attributes(self, 'ImportedRestApi',
                                                        rest_api_id=rest_api_id,
                                                        root_resource_id=rest_api_root_resource_id,
                                                        )
        CfnOutput(self, "ImportedRestApiId",
                  value=imported_api.rest_api_id,
                  description='The ID of this (imported) API Gateway RestApi.',
                  )

        # Fetch dummy Lambda function handler code
        with open("./assets/script.py", encoding="utf8") as fp:
            handler_code = fp.read()
        func_two = lambda_.Function(self, "func_two",
                                    runtime=lambda_.Runtime.PYTHON_3_9,
                                    handler="index.main",
                                    code=lambda_.InlineCode(handler_code),
                                    )

        # Now let's swap out the function with a sandbox!
        new_lambda_backend = LambdaIntegration(func_two)
        resource = imported_api.root.add_resource("v2")

        resource.add_method("GET", api_key_required=True,
                            request_validator=apigateway.RequestValidator(
                                scope=self,
                                rest_api=imported_api,
                                id="api-validator",
                                request_validator_name="api-validator" + '-' + self.stack_name,
                                validate_request_body=True
                            ),
                            integration=new_lambda_backend,
                            request_models={
                                "application/json": request_model
                            })

        rest_api_id = imported_api.rest_api_id
        CfnOutput(self, "RestApiId",
                  value=rest_api_id,
                  description='The ID of this API Gateway RestApi.',
                  )

File: Snippets/ApiGateway/TypeScript/MockIntegration-with-CORS/lab-stack.ts
import * as cdk from 'aws-cdk-lib';
import {CfnOutput} from 'aws-cdk-lib';
import {Construct} from 'constructs';
import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import {MockIntegration, PassthroughBehavior} from 'aws-cdk-lib/aws-apigateway';

export class LabStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        const api = new apigateway.RestApi(this, 'demo-api', {
            restApiName: 'Pretty fly for an API',

            // You can add CORS at the resource-level using addCorsPreflight.
            // https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_apigateway.RestApi.html#defaultcorspreflightoptions
            // defaultCorsPreflightOptions: {
            //     allowOrigins: ['http://localhost:3000'],
            //     allowMethods: apigateway.Cors.DEFAULT_HEADERS
            // }
        });


        //                   _
        //   /\/\   ___   ___| | _____ _ __ _   _
        //  /    \ / _ \ / __| |/ / _ \ '__| | | |
        // / /\/\ \ (_) | (__|   <  __/ |  | |_| |
        // \/    \/\___/ \___|_|\_\___|_|   \__, |
        //                                  |___/

        const integration = new MockIntegration({
            integrationResponses: [
                {statusCode: '200'},
            ],
            passthroughBehavior: PassthroughBehavior.NEVER,
            requestTemplates: {
                'application/json': '{ "statusCode": 200 }',
            },
        })

        let resource = api.root.addResource('mock');

        resource.addMethod('GET', integration, {
            methodResponses: [
                {statusCode: '200'},
            ],
        })

        //  _____ _              ___   ___  __  __
        // /__   \ |__   ___    / __\ /___\/__\/ _\
        //   / /\/ '_ \ / _ \  / /   //  // \//\ \
        //  / /  | | | |  __/ / /___/ \_// _  \_\ \
        //  \/   |_| |_|\___| \____/\___/\/ \_/\__/
        //
        let cors_preflight_options = {
            allowOrigins: ['https://mock.amazon.com'],
            allowMethods: ['GET', 'PUT']
        }

        // Enable CORS
        resource.addCorsPreflight(cors_preflight_options)


        //   ___       _               _
        //   /___\_   _| |_ _ __  _   _| |_ ___
        //  //  // | | | __| '_ \| | | | __/ __|
        // / \_//| |_| | |_| |_) | |_| | |_\__ \
        // \___/  \__,_|\__| .__/ \__,_|\__|___/
        //                 |_|

        new CfnOutput(this, 'RestApiName', {
            description: 'A human friendly name for this Rest API.',
            value: api.restApiName,
        })
        new CfnOutput(this, 'RestApiId', {
            description: 'The ID of this API Gateway RestApi.',
            value: api.restApiId,
        })
        new CfnOutput(this, 'DemoResourceId', {
            description: 'The resource ID of the demo resource.',
            value: resource.resourceId,
        })
    }
}
